### PR TITLE
[GPU] Added several shape agnostic ref kernels

### PR DESCRIPTION
**Pull request by vladimir-paramuzov**
### Details:
 - More shape agnostic ref kernels
 - Should be merged after #13661

### Tickets:
 - *95104*

### CODE REVIEW

File: src/plugins/intel_gpu/src/kernel_selector/kernels/softmax/softmax_kernel_base.cpp [hunk 1]
```diff
@@ -60,13 +59,47 @@ KernelsData SoftmaxKernelBase::GetCommonKernelsData(const Params& params, const
...
-    auto dispatchData = SetDefault(orgParams, options);
+    kd.update_dispatch_data_func = [this](const Params& params, KernelData& kd) {
+        const auto& prim_params = static_cast<const softmax_params&>(params);
+        auto dispatchData = SetDefault(prim_params);
+        OPENVINO_ASSERT(kd.kernels.size() == 1, "[GPU] Invalid kernels size for update dispatch data func");
+        kd.kernels[0].params.workGroups.global = dispatchData.gws;
+        kd.kernels[0].params.workGroups.local = dispatchData.lws;
+        kd.internalBufferSizes.clear();
+        kd.internalBufferSizes.push_back(prim_params.inputs[0].PhysicalSizeInBytes());
+        kd.internalBufferDataType = Datatype::F32;
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/softmax/softmax_kernel_base.cpp:70]: `GetAccumulatorType()` may return fp16 either fp32 data type

File: src/plugins/intel_gpu/tests/test_cases/softmax_gpu_test.cpp [hunk 2]
```diff
@@ -300,6 +302,10 @@ TEST(softmax_gpu_bfyx_f32, normalize_f) {
 
     auto output = outputs.at("softmax").get_memory();
     cldnn::mem_lock<float> output_ptr(output, get_test_stream());
+    for (size_t i = 0; i < output->count(); i++) {
+        std::cerr << "i = " << i << " v = " << output_ptr[i] << std::endl;
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/tests/test_cases/softmax_gpu_test.cpp:306]: Please delete debug code

File: src/plugins/intel_gpu/tests/test_cases/softmax_gpu_test.cpp [hunk 3]
```diff
@@ -976,3 +982,94 @@ INSTANTIATE_TEST_SUITE_P(softmax_gpu_formats_test_f16_3d,
...
+    cldnn::mem_lock<float> output_ptr(output, get_test_stream());
+    float out_buffer[buf_size];
+    for (uint32_t i = 0; i < buf_size; i++) {
+        out_buffer[i] = output_ptr[i];
+    }
+
+    float temp_max = 0;
+    float expected_sum = 1.0f;
+    int max_value_buffer_index = 0;
+    for (size_t i = 0; i < output->count(); i++) {
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/tests/test_cases/softmax_gpu_test.cpp:1049]: Debug code

File: src/plugins/intel_gpu/tests/test_cases/softmax_gpu_test.cpp [hunk 4]
```diff
@@ -976,3 +982,94 @@ INSTANTIATE_TEST_SUITE_P(softmax_gpu_formats_test_f16_3d,
...
+                    int index = i * f * x * y +
+                                l * x * y +
+                                j * x +
+                                k;
+                    if (out_buffer[index] >= temp_max) {
+                        temp_max = out_buffer[index];
+                    }
+                    sum += out_buffer[index];
+                }
+                EXPECT_EQ(true, are_equal(temp_max, expected_max_values[max_value_buffer_index]));
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/tests/test_cases/softmax_gpu_test.cpp:1066]: `ASSERT_TRUE(are_equal(temp_max, expected_max_values[max_value_buffer_index]);`

File: src/plugins/intel_gpu/src/graph/graph_optimizer/compile_graph.cpp [hunk 5]
```diff
@@ -36,6 +37,11 @@ void compile_graph::run(program& p) {
         bool can_select_impl = !node->is_type<data>() &&
                                !(node->is_type<mutable_data>() && node->get_dependencies().empty()) &&
                                (!node->is_dynamic() || node->type()->does_dynamic_implementation_exist(*node));
+
+        // TODO: Remove this WA once we have shape agnostic reshape kernel
+        if (node->is_type<reshape>() && node->is_dynamic() && !node->can_be_optimized())
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/compile_graph.cpp:42]: Dont we need to optimize out reahape in handle_reshape?

**Comment by yeonbok** [comment]
As I mentioned offline, this causes additional accuracy issue for dynamic shape. Please check that..

File: src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_base.cpp [hunk 6]
```diff
@@ -17,6 +17,11 @@ bool PermuteKernelBase::Validate(const Params& p, const optional_params& o) cons
             return false;
         }
     }
+
+    auto supported_dyn_layouts = {DataLayout::bfyx, DataLayout::bfzyx, DataLayout::bfwzyx};
```
[the remark this answers is not in the document]
Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_base.cpp:21]: We can think about unified approach for supported dynamic params check, but not in scope of this PR

File: src/plugins/intel_gpu/src/graph/graph_optimizer/compile_graph.cpp (hunk 5 above)
Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/compile_graph.cpp:42]: Probably no. handle reshape is done on early graph optimization stages, so we can't be sure that optimized flag won't be invalidated after fusions or some other passes.

**Review by sshlyapn — APPROVED** [review]

**Pull request merged**